### PR TITLE
Delete schedule of renamed/deleted flows on project upload.

### PR DESCRIPTION
**Pull request by ypadron-in**
When a new project version is uploaded existing schedules of renamed/deleted flows must be removed.

### CODE REVIEW

**Comment by djaiswal83** [comment]
Thanks for fixing this. Is it possible to add a test for it?

**Comment by djaiswal83** [comment]
There is conflict in the code. Please resolve.

File: azkaban-web-server/src/main/java/azkaban/webapp/servlet/ProjectManagerServlet.java [hunk 1]
```diff
@@ -1728,6 +1729,28 @@ private void ajaxHandleUpload(final HttpServletRequest req, final HttpServletRes
...
         }
 
+        // remove schedule of renamed/deleted flows
+        final Set<String> flowNameList = new HashSet<>();
+        for (final Flow flow : project.getFlows()) {
+          flowNameList.add(flow.getId());
+        }
+
+        try {
+          for (final Schedule schedule : this.scheduleManager.getSchedules()) {
```
Comment (jamiesjc) [inline azkaban-web-server/src/main/java/azkaban/webapp/servlet/ProjectManagerServlet.java]: In some clusters, we have thousands of schedules. Will this check impact the uploading time for users?

**Comment by jamiesjc** [comment]
One minor suggestion: the `ajaxHandleUpload` method is already huge. Would it be nice to move some functionalities out into smaller pieces?

File: azkaban-web-server/src/main/java/azkaban/webapp/servlet/ProjectManagerServlet.java (hunk 1 above)
Comment (djaiswal83) [reply azkaban-web-server/src/main/java/azkaban/webapp/servlet/ProjectManagerServlet.java]: Maybe we can measure it and test it out on a dev cluster? If this comes out to be a real problem, then we can maintain a set of scheduleIds in Project class. Then we only look up schedules for the given project.
ScheduleManager already maintain a map of scheduleID to Schedule which can be used to remove the schedule.

Comment (ypadron-in) [reply azkaban-web-server/src/main/java/azkaban/webapp/servlet/ProjectManagerServlet.java]: I don't think the check will add considerable latency. I tested locally with a List of 1 million objects containing an Integer and the time it took to iterate and compare values is under 15 milliseconds.

File: azkaban-web-server/src/main/java/azkaban/webapp/servlet/ProjectManagerServlet.java [hunk 2]
```diff
@@ -1774,144 +1784,177 @@ private void ajaxHandleUpload(final HttpServletRequest req, final HttpServletRes
...
+  static void removeScheduleOfDeletedFlows(final Project project,
+      final ScheduleManager scheduleManager, final Consumer<Schedule> onDeletedSchedule)
+      throws ScheduleManagerException {
+    final Set<String> flowNameList = project.getFlows().stream().map(f -> f.getId()).collect(
+        Collectors.toSet());
 
-        if (this.enableQuartz) {
-          this.scheduler.schedule(project, user.getUserId());
-        }
+    for (final Schedule schedule : scheduleManager.getSchedules()) {
```
Comment (edwinalu) [inline azkaban-web-server/src/main/java/azkaban/webapp/servlet/ProjectManagerServlet.java]: Would it be possible to get schedules by project, to limit the number of flows fetched?

File: azkaban-web-server/src/main/java/azkaban/webapp/servlet/ProjectManagerServlet.java [hunk 3]
```diff
@@ -1774,144 +1784,177 @@ private void ajaxHandleUpload(final HttpServletRequest req, final HttpServletRes
...
+            schedule.getScheduleId(), schedule.getFlowName(), schedule.getProjectName());
+        this.projectManager.postProjectEvent(project, EventType.SCHEDULE, "azkaban",
+            "Schedule " + schedule.toString() + " has been removed.");
+      });
+
+      registerErrorsAndWarningsFromValidationReport(resp, ret, reports);
+    } catch (final Exception e) {
+      logger.info("Installation Failed.", e);
+      String error = e.getMessage();
+      if (error.length() > 512) {
```
Comment (edwinalu) [inline azkaban-web-server/src/main/java/azkaban/webapp/servlet/ProjectManagerServlet.java]: Please add some constants for 512 and 500.

File: azkaban-web-server/src/main/java/azkaban/webapp/servlet/ProjectManagerServlet.java [hunk 4]
```diff
@@ -1774,144 +1784,177 @@ private void ajaxHandleUpload(final HttpServletRequest req, final HttpServletRes
...
-
-      logger.info("Upload: project " + projectName + " version is " + project.getVersion()
-          + ", reference is " + System.identityHashCode(project));
-      ret.put("version", String.valueOf(project.getVersion()));
+    }
+    if (errorMsgs.length() > 0) {
+      // If putting more than 4000 characters in the cookie, the entire message will somehow
+      // get discarded.
+      registerError(ret,
+          errorMsgs.length() > 4000 ? errorMsgs.substring(0, 4000) : errorMsgs.toString(), resp,
```
Comment (edwinalu) [inline azkaban-web-server/src/main/java/azkaban/webapp/servlet/ProjectManagerServlet.java]: Is there a reason why this is 4000 and not 512? Can this be consistent?

File: azkaban-web-server/src/test/java/azkaban/webapp/servlet/ProjectManagerServletTest.java [hunk 5]
```diff
@@ -0,0 +1,60 @@
+package azkaban.webapp.servlet;
```
Comment (edwinalu) [inline azkaban-web-server/src/test/java/azkaban/webapp/servlet/ProjectManagerServletTest.java]: Please add the copyright info.

File: azkaban-web-server/src/main/java/azkaban/webapp/servlet/ProjectManagerServlet.java (hunk 2 above)
Comment (ypadron-in) [reply azkaban-web-server/src/main/java/azkaban/webapp/servlet/ProjectManagerServlet.java]: It looks like there is no direct relationship between Project and Schedule. The list of schedules is coming from memory and Im filtering out by project first to reduce work. Creating a getSchedulesByProject() method in ScheduleManager class will do essentially the same thing I'm doing.

File: azkaban-web-server/src/test/java/azkaban/webapp/servlet/ProjectManagerServletTest.java (hunk 5 above)
Comment (ypadron-in) [reply azkaban-web-server/src/test/java/azkaban/webapp/servlet/ProjectManagerServletTest.java]: done, thanks.

File: azkaban-web-server/src/main/java/azkaban/webapp/servlet/ProjectManagerServlet.java (hunk 3 above)
Comment (ypadron-in) [reply azkaban-web-server/src/main/java/azkaban/webapp/servlet/ProjectManagerServlet.java]: Using constants for HTTP status code now, thanks.

File: azkaban-web-server/src/main/java/azkaban/webapp/servlet/ProjectManagerServlet.java (hunk 4 above)
Comment (ypadron-in) [reply azkaban-web-server/src/main/java/azkaban/webapp/servlet/ProjectManagerServlet.java]: I'm only moving the code around to reduce the size of ajaxHandleUpload(). I don't know what these numbers: 512 and 4000 represent.

Comment (djaiswal83) [reply azkaban-web-server/src/main/java/azkaban/webapp/servlet/ProjectManagerServlet.java]: I think this is a good chance to understand why the numbers are different. Can you please find out?

Comment (ypadron-in) [reply azkaban-web-server/src/main/java/azkaban/webapp/servlet/ProjectManagerServlet.java]: 512 limits the size of an unexpected error message while processing the request and seems a pretty arbitrary choice.
4000 limits the size of project validation messages, apparently it has something to do with browser cookies, but if that's the case the right place to do the trimming is when it's actually stored in a cookie, not here.
This is the only place where these numbers are used and we wouldn't be gaining much by making this change. Given our current priorities I would suggest to leave it as it is and revisit this when we have a better motivation.

File: azkaban-web-server/src/main/java/azkaban/webapp/servlet/ProjectManagerServlet.java [hunk 6]
```diff
@@ -1774,144 +1785,178 @@ private void ajaxHandleUpload(final HttpServletRequest req, final HttpServletRes
...
+  private Project validateUploadAndGetProject(final HttpServletResponse resp,
+      final Map<String, String> ret, final User user, final String projectName) {
+    if (projectName == null || projectName.isEmpty()) {
+      registerError(ret, "No project name found.", resp, HttpServletResponse.SC_BAD_REQUEST);
+      return null;
+    }
+    final Project project = this.projectManager.getProject(projectName);
+    if (project == null || !project.isActive()) {
+      final String failureCause = (project == null) ? "doesn't exist." : "was already removed.";
+      registerError(ret, "Installation Failed. Project '" + projectName + " "
```
Comment (li-ygerchikov) [inline azkaban-web-server/src/main/java/azkaban/webapp/servlet/ProjectManagerServlet.java:1862]: Missing the closing single-quota. Not worth its own PR, but if we are ever changing this, or anything close to it...

**Pull request merged**